### ISSUE TITLE
`<ranges>`: ICE with std::views::join_with

### ISSUE THREAD

**Issue by cpplearner** [StephanTLavavej/STL#17]
# Basic Info
* [x] I have read the rules for the [Standard Library Modules Bug Bash][bug-bash-url].

* [x] I have pulled and rebuilt the `bug-bash` branch in the last 24 hours.

* My current commit, as printed by `git rev-parse HEAD`, is: 6c96fe702ab23df846cca2110bc8dc08e2aa6d3a

* This bug is:
ICE-on-valid

# Command-Line Test Case
My test case that works with classic includes but fails with named modules:

```
D:\bug-bash>type test.cpp
```
```cpp
#ifdef USE_CLASSIC_INCLUDES
    #include <array>
    #include <ranges>
#else
    import std;
#endif

int main() {
    std::array<std::array<char, 10>, 10>{} | std::views::join_with(' ');
}
```
```
D:\bug-bash>cl /EHsc /nologo /W4 /std:c++latest /MTd /c /FoNUL /DUSE_CLASSIC_INCLUDES test.cpp
test.cpp

D:\bug-bash>cl /EHsc /nologo /W4 /std:c++latest /MTd /c /FoNUL test.cpp
test.cpp
D:\bug-bash\out\x64\out\inc\ranges(3883): fatal error C1001: Internal compiler error.
(compiler file 'msc1.cpp', line 1576)
 To work around this problem, try simplifying or changing the program near the locations listed above.
If possible please provide a repro here: https://developercommunity.visualstudio.com
Please choose the Technical Support command on the Visual C++
 Help menu, or open the Technical Support help file for more information
INTERNAL COMPILER ERROR in 'D:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31627\bin\HostX64\x64\cl.exe'
    Please choose the Technical Support command on the Visual C++
    Help menu, or open the Technical Support help file for more information
```

# Additional Context
None.

[bug-bash-url]: https://github.com/StephanTLavavej/STL/wiki/Standard-Library-Modules-Bug-Bash

**Comment by StephanTLavavej** [comment]
Thanks for the excellent test case! I'll report this to the compiler team later today.

**Comment by StephanTLavavej** [comment]
Confirmed, still repros with 17.3 Preview 5. Reported as internal [VSO-1582910](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1582910) "Standard Library Modules: ICE with `std::views::join_with`".

**Comment by StephanTLavavej** [comment]
A compiler fix for this ICE is undergoing review!

**Comment by StephanTLavavej** [comment]
Compiler fix will ship in 17.4 Preview 3.
